Update the chapter 'Composing large statements' v4.4 (#152)
neo4j/docs-getting-started: modules/ROOT/pages/cypher-intro/large-statements.adoc

[[cypher-intro-large-statements-with]]
== WITH

> In Cypher, you are able to chain fragments of statements together, similar to how it is done within a data-flow pipeline.

Reviewer: If you must use a personal pronoun, I'd suggest something like:
_In Cypher, you can chain fragments..._

Each fragment works on the output from the previous one, and its results can feed into the next one.
_Only_ columns declared in the `WITH` clause are available in subsequent query parts.

The `WITH` clause is used to combine the individual parts and declare which data flows from one to the other.
`WITH` is similar to the `RETURN` clause.
The difference is that the `WITH` clause does not finish the query, but prepares the input for the next part.
Expressions, aggregations, ordering, and pagination can be used in the same way as in the `RETURN` clause.
The only difference is all columns must be aliased.

> In the example below, we collect the movies someone appeared in and then filter out those which appear in only one movie.

Reviewer: ```suggestion
In the example below, collect the movies someone appeared in and then filter out those which appear in only one movie.
```

+-------------------------------------------------------------+
----

Using the `WITH` clause, you can pass values from one section of a query to another.
This allows you to perform some intermediate calculations or operations within your query to use later.

> Take a look at the following examples based on this dataset:

Reviewer: ```suggestion
The following examples are based on this dataset:
```

image:people-technologies-graph-arr.svg[role='popup-link']

> Specifying in the `WITH` clause variables you want to use later, you can apply this functionality in various ways (for example, count, collect, filter, or limit results).

Reviewer: ```suggestion
In the `WITH` clause, specify the variables you want to use later.
This functionality can be applied in various ways (for example, count, collect, filter, or limit results).
```

> Here you can find a couple of examples, including a cleaner version of the `size()` query from above (see xref:cypher-intro/results.adoc#aggregate-size[Counting values in a list]).

Reviewer: ```suggestion
A few examples of this follows, including a cleaner version of the `size()` query from above (see xref:cypher-intro/results.adoc#aggregate-size[Counting values in a list]).
```

| 'Joe'       | ['Query Languages']                           |
| 'Melissa'   | ['Query Languages', 'Data ETL', 'Graphs']     |
+-------------------------------------------------------------+
----

[source, cypher]
----
//Query2: find number of friends who have other friends
MATCH (p:Person)-[:IS_FRIENDS_WITH]->(friend:Person)
WITH p, collect(friend.name) AS friendsList, size((friend)-[:IS_FRIENDS_WITH]-(:Person)) AS numberOfFoFs
WHERE numberOfFoFs > 1
RETURN p.name, friendsList, numberOfFoFs;
----

*Query2 results:*

[queryresult]
----
Rows: 3

+-----------------------------------------------------------------+
| p.name     | friendList                        | numberOfFoFs   |
+-----------------------------------------------------------------+
| 'Joe'      | ['Mark']                          | 2              |
| 'John'     | ['Sally']                         | 2              |
| 'Jennifer' | ['Sally', 'John', 'Ann', 'Mark']  | 2              |
+-----------------------------------------------------------------+
----

In the first query, the `Person` name and a collected list of the `Technology` types are declared.
Only those items can be referenced in the `RETURN` clause.

> You cannot use the relationship (`r`) or the `Person` twitter because those values were not passed along.

Reviewer: ```suggestion
The relationship (`r`) or the `Person` twitter cannot be used because those values were not passed along.
```

> In the second query, you can reference `p` and any of its properties (`name`, `yearsExperience`, `twitter`), the collection of friends (as a whole, not each value), and the number of friend-of-friends.

Reviewer: ```suggestion
In the second query, only `p` and any of its properties (`name`, `yearsExperience`, `twitter`), the collection of friends (as a whole, not each value), and the number of friend-of-friends can be referenced.
```

> Since you passed those values in the `WITH` clause, you can use those in the `WHERE` or `RETURN` clauses.

Reviewer: ```suggestion
Since those values were passed along in the `WITH` clause, they can be used in the `WHERE` or `RETURN` clauses.
```

`WITH` requires all values passed to have a variable (if they do not already have one).

> The `Person` nodes were given a variable (`p`) in the `MATCH` clause, so you do not need to assign a variable there.

Reviewer: ```suggestion
The `Person` nodes were given a variable (`p`) in the `MATCH` clause, so no variable needs to be assigned there.
```
